### ISSUE TITLE
UtBS - formation ability - misleading defense preview

### ISSUE THREAD

**Issue by doofus-01** [wesnoth/wesnoth#4917]
### Game and System Information
 - What version of the game are you running?
1.15.3+dev 
 - Where did you download it from? (Steam, Mac App Store, our website, built from source, etc.)
source
 - If you're using a custom build, what compiler did you use, and what commit did you build at?
g++/scons
44be67b6fa0
 - What OS are you running?
Linux

### Describe the bug
Formation ability should give a 10% boost to defense, and in the attack dialog it does, but the on-map defense preview doesn't reflect that.  See attached image, it shows the on-map 40% defense, but the attack dialog shows 50% CTH for the enemy (like it should).

### To Reproduce
Play UtBS

### Expected behavior
The on-map and attack dialog (and actual) defense numbers should match
### Screenshots
![Screenshot_20200523](https://user-images.githubusercontent.com/13510196/82744530-a45f8000-9d2e-11ea-95f4-ca78638ddd0c.png)

**Comment by gfgtdf** [comment]
Accroding to the comment

https://github.com/wesnoth/wesnoth/blob/master/src/units/unit.cpp#L1652

, we explciitly don't want to take abilities into account when calculating the defense for a unit on a tile, also for performance reasons.

**Comment by doofus-01** [comment]
So I guess it's a deeper problem, and probably won't be fixed, but it's still a bug.

**Comment by soliton-** [comment]
You could try if activating that code (remove the `#if 0` and `#endif` lines) does fix it and if you notice performance degradation.

I can't quite imagine why it would be an issue for the mouse over which should only call this function once you move to a different hex. However that code is probably used in different contexts as well. Perhaps we just need two versions of this function and call the accurate one in UI context and the simplified one in AI or whatever other context it is called where performance is more important.

**Comment by doofus-01** [comment]
OK, I tried that, and did not notice any change.  I didn't test AI performance though, since it didn't fix the incorrect defense number.

**Comment by soliton-** [comment]
Oh well. I guess that's because of the "doesn't take into account target locations" part. Not sure how easy it'd be to add that.

**Comment by gfgtdf** [comment]
There are more things that the game engine at that part doesn't know about for example whether the ability is active also depends on which weapon is chosen, which weapon the attacker chose. In this case of the formation ability the abiltiy is independent on that but for a general solution we would either need to think about whether there is a good way on how to solve this, and what values to show for a defense abiltiy that only effect certain weapons ore against enemies.

And before we got a with a solution that might look better in this particular case its clearly better to go with a easy to understand "abilities changes not included" for the path-display defense value.

**Comment by doofus-01** [comment]
I don't think the player would expect something that depends on weapons used to be shown outside the attack dialog.  For something like this, it depends on map position, just like any terrain difference, and would seem to be the whole point of that displayed number.  Displaying a wrong value is not clearly better.

**Comment by soliton-** [comment]
I agree that it's not clearer to show the wrong value if there is no obvious reason for it. For an ability that depends on some weapon choice it seems quite obvious that without such choice the game cannot know what value to display so it must show a value without considering that ability.

The current situation is only easy to understand if you know the code.

**Comment by gfgtdf** [comment]
> For an ability that depends on some weapon choice it seems quite obvious that without such choice the game cannot know what value to display so it must show a value without considering that ability.

Not really, what if you have an ability with a [filter_weapon] that happens to match all weapons the unit has? Do you show the modified value or not? This is a nontrivial question since from the players perspective he usually doesn't know anything at all about the [filter_weapon], on the other hand, computing the ability for every weapon might not only be slow but also unreliable.

> The current situation is only easy to understand if you know the code.

I disagree here, i don't think its hard to understand that defense values depend on terrain and on abilities/weaponspecials. Currently we for example also don't show the defense as 30% if all enemy attacks of every enemy unit in range have the magical special, which would effectively lead to the unit having always 30% defense 


I still really think a simple "defense depends on terrain and on weapon specials" is in the end easier to understand for the users than whatever solution someone might propose for this issue, because it would most likely end up showing the modified terrain defense for some types of abilities and for other types of abilities the basic terrain defense with no patters that would make sense from a users perspective.


Furthermore in this example of the first post the unit has only 50% defense there as long as the other unit doesn't move. a player that reads the 50% might aswell be not aware of it and assume the unit has 50% defense on that terrain after he moved the other unit.

So, if someone still want to implement something here, make it obvious that the value shown is a modified terrain defense and not the base terrains defense. For example by writing the numbers fat or something like that.

**Comment by soliton-** [comment]
> Not really, what if you have an ability with a [filter_weapon] that happens to match all weapons the unit has? Do you show the modified value or not? This is a nontrivial question since from the players perspective he usually doesn't know anything at all about the [filter_weapon], on the other hand, computing the ability for every weapon might not only be slow but also unreliable.

If the value does not depend on the weapon choice because all weapons match or all produce the same value anyway then we should show the correct amount.

The player should definitely know about the `[filter_weapon]` from some description. What situations do you have in mind where it would be a good idea to not explain how an ability works?

Can you explain what's unreliable about the computation?

I cannot really say much about performance but any computation that is only triggered by user interaction has to be really slow to be noticable.

> I disagree here, i don't think its hard to understand that defense values depend on terrain and on abilities/weaponspecials.

So you're agreeing that we should show the correct value based on abilities/weapon specials? I'm a little confused.

To perhaps clarify what I meant: Unless you know the code that shows you abilities are not taken into account I see no reason why a player would expect that to be the case. If that is explained somewhere in help or wherever I agree that it wouldn't be difficult to explain or understand. (If players will find and read that explanation is yet a different matter.)

> Currently we for example also don't show the defense as 30% if all enemy attacks of every enemy unit in range have the magical special, which would effectively lead to the unit having always 30% defense

I think that argument is quite far-fetched but sure if you really think that is what players expect to see then we could make it so.

> I still really think a simple "defense depends on terrain and on weapon specials" is in the end easier to understand for the users than whatever solution someone might propose for this issue, because it would most likely end up showing the modified terrain defense for some types of abilities and for other types of abilities the basic terrain defense with no patters that would make sense from a users perspective.

Doesn't seem that way to me but you surely know more about that so I'll take your word for it.

Still confused why you say "defense depends on terrain and on weapon specials" as if someone disagrees with that.

> Furthermore in this example of the first post the unit has only 50% defense there as long as the other unit doesn't move. a player that reads the 50% might aswell be not aware of it and assume the unit has 50% defense on that terrain after he moved the other unit.

> So, if someone still want to implement something here, make it obvious that the value shown is a modified terrain defense and not the base terrains defense. For example by writing the numbers fat or something like that.

Indeed a visual clue that it is not just the plain terrain defense or maybe even showing the modified and the base value together in some way sounds like a good idea.

**Comment by gfgtdf** [comment]
> If the value does not depend on the weapon choice because all weapons match or all produce the same value anyway then we should show the correct amount.

my correct you mean including the abilities right? so the value that is shown in the attack dialog.

> So you're agreeing that we should show the correct value based on abilities/weapon specials? I'm a little confused.
To perhaps clarify what I meant: Unless you know the code that shows you abilities are not taken into account I see no reason why a player would expect that to be the case. If that is explained somewhere in help or wherever I agree that it wouldn't be difficult to explain or understand. (If players will find and read that explanation is yet a different matter.)
Still confused why you say "defense depends on terrain and on weapon specials" as if someone disagrees with that.

Hmm no, i mean as in showing the base terrain value, and assume that he player knows that the actual value used in combat is additionaly infulenced  by weapon specials or abilities.


> The player should definitely know about the [filter_weapon] from some description. What situations do you have in mind where it would be a good idea to not explain how an ability works?

No i never said there is a case where the player should not know how an ability works, what i was saying is: if the code would only show the modified value for abilities without for example filter_weapon, then it'd be couintertinutitive, since for example even an empty [filter_weapon] which might appear for teachnical reason for the code that generated the ability or whatever,  since the user would not know about filter_weapon here.


> I cannot really say much about performance but any computation that is only triggered by user interaction has to be really slow to be noticable.


I don't know when this code is called exactly, its for example quite thinkable that this code is called for every tile around unit during shortest path computation, to prefer paths with better defense. I only took the comment for performance concerns in the code and if we do it for every weapon we will probably triple the amount of times we compute this value in average. If we assume it was 'slow' before it will then be even slower. (oh and keep in mind that we also have to take into consideration 'the no retaliation weapon avilable' case, that is the 'null' weapon pasicially, (probably not important  here just a nice fact)). And if we take it further like computing it 'for every enemy unit in range' it will probably become actually slow.

The good new is however, the filter code has quite some space left for optimisation, in face someone wants to do it :), for exampel we can store use filter objects like unit_filter instead of just config objects in the attack.


> Doesn't seem that way to me but you surely know more about that so I'll take your word for it.

> Can you explain what's unreliable about the computation?

Afaik currently the code that does the computations always assume an actual attack context, that is a particular an opponent, filters might use variables like opponent.level in filter_self, making  harder for the engine to detect that the weapon specials does actually depend on the opponent.

It's mostly that i think are many corner cases, for example a naive 'solution' to this problem would probably show the 'modified' value on all terrains for a 'this unit has 10% defense if it has full movement this turn' ability since when evaluating the filter the unit still has full movement.

You could make an ability that changes defense for all weapons by 20% except for one particular weapon where it increases by 30%, do you want to show a 20% modified value or the base terrain value?

You can have an ability that changes defense only during your turn, it might be unhandly to show the modified value if we are mostly interested in it for building our defense line.

You can have an ability that uses lua_function in a filter where the engine has have no idea at all what it does.

And i think umc authors would find more corner cases.

**Comment by AI0867** [comment]
> It's mostly that i think are many corner cases, for example a naive 'solution' to this problem would probably show the 'modified' value on all terrains for a 'this unit has 10% defense if it has full movement this turn' ability since when evaluating the filter the unit still has full movement.
> 
> You could make an ability that changes defense for all weapons by 20% except for one particular weapon where it increases by 30%, do you want to show a 20% modified value or the base terrain value?
> 
> You can have an ability that changes defense only during your turn, it might be unhandly to show the modified value if we are mostly interested in it for building our defense line.
> 
> You can have an ability that uses lua_function in a filter where the engine has have no idea at all what it does.
> 
> And i think umc authors would find more corner cases.

Easy. Here are various cases where the unit can't defend itself using an attack, and therefore can't use the weapon specials.

All the unit's weapons add 10% defense, but they're all melee weapons and the enemy attacks using a ranged unit.

All the unit's weapons add 10% defense, but some ability has disabled its attacks.

All the unit's weapons add 10% defense, and it has both melee and ranged weapons, but it's used in combination with an era that adds a third range.

**Comment by doofus-01** [comment]
These examples all look like they are tied to weapon choice (even if there isn't much of a choice), which makes sense to look for in the attack dialog.  If the modifier is based on map placement, the number on the map matters.  The player doesn't know or care how the underlying filter is implemented (I haven't really looked either).  The presentation matters, what's this all for otherwise?

At the end of the day, this "formation" ability is a mainline feature now, not some UMC cleverness, so waving it off because fixing it is a slippery slope is pretty weak.  Maybe "formation" shouldn't have been put in mainline, maybe the bug will never get fixed.  But it is an open bug.

**Comment by AI0867** [comment]
True, abilities would make some degree of sense, but only if their effects are independent of the other units on the map. (move away/lose a different unit, and suddenly the defense drops? I didn't expect that!)

While I can still imagine such abilities (better defense when you have impassable terrain at your side, for example), most of them would be better implemented as simple changes to unit movetype.

I haven't looked at the related code in a while, but would excluding all units from the calculation speed things up? I really don't know how simple or complex it would be to implement this though. If it's too complex, I would recommend just dropping the whole idea.

From a UI perspective, it would be nice to have some sort of indication that "in this configuration, this ability or other has this effect", but that would be even more work.
